### PR TITLE
Feature/14 create pause menu

### LINKED ISSUE #14: Create Pause Menu
**Short Description**
When pressing <ESC> in the game screen a pause menu with 3 Buttons will pop up and nothing else will be rendered. 

**Acceptance**
- When pressing <ESC> in the game screen a pause menu with 3 Buttons will pop up 
- Buttons are "resume" "save" "save & exit"
- resume returns to the game
- save just logs "saving...'
- "save & exit" calls the save functionallity and then exits the game. 

**Implementation**
- https://libgdx.com/wiki/graphics/2d/scene2d/scene2d-ui
- seperate it from the by using a different viewport 
- https://libgdx.com/wiki/graphics/2d/scene2d/scene2d
- https://libgdx.com/wiki/graphics/viewports

### PR DESCRIPTION
**Pull request by hearnoli**
Fixes #14

## Checklist (for code changes)

- [x] added docs
- [ ] added tests

## Short description

Added a pause menu with the changes described in the issue: Resume, Save, Save and Exit

### CODE REVIEW

**Comment by BetaTH13** [comment]
Can we load the Skins via the assetmanager? 
@TheBromo Regarding the Mutiplexer, do we need that RSAInputAdapter you made on one of those demos branches?

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ * Shows a pause screen with options to resume, save, or save and exit the game.
+ */
+public class PauseScreen implements Screen {
+
+    private FrontierGame frontierGame;
+    private Stage stage;
+    private Skin skin;
+
+    public PauseScreen(FrontierGame frontierGame) {
```
Comment (TheBromo) [inline frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:27]:
```suggestion
    public PauseScreen(FrontierGame frontierGame, GameScreen gameScreen) {
       this.gameScreen = gameScreen;
```


we also need to save the game screen, since we want to open it with the same values again.

**Comment by hearnoli** [comment]
@BetaTH13 Yes, we can load Skins with the AssetManager: assetManager.load("uiskin.json", Skin.class);

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        this.frontierGame = frontierGame;
+
+        this.stage = new Stage(new ScreenViewport(new OrthographicCamera()));
+        this.skin = new Skin(Gdx.files.internal("uiskin.json"));
```
Comment (TheBromo) [inline frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:31]: loading assets has to be done with the assetmanager in  the loading screen..

**Review by TheBromo — CHANGES_REQUESTED** [review]
another point is the loading screen opens the game and not the menu at the moment.

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java (hunk 1 above)
Comment (BetaTH13) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:27]: @TheBromo I am wondering isn't it a good idea to store and load the GameScreen values from the FrontierGame? I am thinking of something similar to PM3 where we where storing all the Information for the game in the gamestate.

**Comment by TheBromo** [comment]
+ the code needs to be formatted. in the root readme is the commmand.

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java (hunk 1 above)
Comment (TheBromo) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:27]: yes this can also be done.

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+public class PauseScreen implements Screen {
+
+    private FrontierGame frontierGame;
+    private Stage stage;
+    private Skin skin;
+
+    public PauseScreen(FrontierGame frontierGame) {
+        this.frontierGame = frontierGame;
+
+        this.stage = new Stage(new ScreenViewport(new OrthographicCamera()));
```
[the remark this answers is not in the document]
Comment (TheBromo) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:30]: one suggestion would be 

```suggestion
        this.stage = new Stage(new ExtendViewport(1920, 1080, new OrthographicCamera()));
```

this results in this on a 4k display

image

**Review by TheBromo — CHANGES_REQUESTED** [review]

**Comment by TheBromo** [comment]
@hearnoli i would also suggest maybe adding a integration test that checks wether pressing the esc key while being in the gamescreen switches to the pause screen.

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (BetaTH13) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:30]: Manuel i think we should in general rework the viewports. In all of the issues we have at the moment the viewport is configured and used differently. I would say we leave them in general out of scope and we have another issue to fix the viewports.

**Comment by BetaTH13** [comment]
@TheBromo For the unit test i made some work arounds. For example i moved almost everything which was in the constructor of GameScreen and PauseScreen into show method. I also added some conditions to the dispose method. I couldn't really come up with another way to test render (pressing ESC) function.

Do you think you think this is oke for now or do we need to change something else?

File: frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java [hunk 4]
```diff
@@ -176,8 +179,12 @@ public void dispose() {
     }
 
     private void updateUI() {
-        gameUi.apply();
-        stage.act();
-        stage.draw();
+        if (gameUi != null) {
```
Comment (TheBromo) [inline frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java:182]: why do we need to null check this?

File: frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java [hunk 5]
```diff
@@ -176,8 +179,12 @@ public void dispose() {
...
+            gameUi.apply();
+        }
+        if (stage != null) {
```
Comment (TheBromo) [inline frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java:185]: same here: why do we need to null check this?

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            }
+        );
+
+        table.add(resumeButton).fillX().pad(10);
+        table.row();
+        table.add(saveButton).fillX().pad(10);
+        table.row();
+        table.add(saveExitButton).fillX().pad(10);
+
+        Gdx.input.setInputProcessor(new InputMultiplexer(stage));
```
Comment (TheBromo) [inline frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:81]: can you move this to an input class an call it something like `PauseScreenUi.java` or something similar

File: frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+import com.badlogic.gdx.scenes.scene2d.ui.Table;
+import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
+import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
+import com.badlogic.gdx.utils.viewport.ExtendViewport;
+import com.zhaw.frontier.FrontierGame;
+
+/**
+ * Shows a pause screen with options to resume, save, or save and exit the game.
+ */
+public class PauseScreen implements Screen {
```
Comment (TheBromo) [inline frontier/core/src/main/java/com/zhaw/frontier/screens/PauseScreen.java:21]:
```suggestion
public class PauseScreen extends ScreenAdapter {
```
so you can remove the empty methods

**Review by TheBromo — CHANGES_REQUESTED** [review]
just a few changes and questions than it is ok to merge

File: frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java (hunk 5 above)
Comment (BetaTH13) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java:185]: This was mainly added so that the unit test would. When we call the switchscreen we dispose of resources right? In the unit tests we are not executing the show/create function, so some elements like stage, gameui are null. The issue we had with the unit test was ofc the spritebatch again. 

We can talk about it together later if i need to clarify something more.

Comment (TheBromo) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java:185]: can we not mock them?

Comment (BetaTH13) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java:185]: I am not sure what the limits would be for the individual objects, but we would need to create getters and setter for testing, currently the stage and gameUi are not accessible from outside. I was not sure what in this case is the ideal solution. Creating getters and setter for unit tests is not really a clean way to do it, how i did it is also not a good approach i basically changed the code for the unit test.

Comment (TheBromo) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java:193]: i personally would prefer to just add `@Setter` for testing purposes than this null checks.

Comment (TheBromo) [reply frontier/core/src/main/java/com/zhaw/frontier/screens/GameScreen.java:193]: or you add a helper class to the tests:
```java
package com.zhaw.frontier.screens;

import com.badlogic.gdx.utils.viewport.ExtendViewport;
import com.badlogic.gdx.utils.viewport.ScreenViewport;

public class GameScreenTestUtils {

    public static void setGameWorldView(GameScreen screen, ExtendViewport view) {
        setPrivateField(screen, "gameWorldView", view);
    }

    public static void setGameUi(GameScreen screen, ScreenViewport ui) {
        setPrivateField(screen, "gameUi", ui);
    }

    private static void setPrivateField(Object obj, String fieldName, Object value) {
        try {
            var field = obj.getClass().getDeclaredField(fieldName);
            field.setAccessible(true);
            field.set(obj, value);
        } catch (Exception e) {
            throw new RuntimeException("Failed to set field: " + fieldName, e);
        }
    }
}

```

then you can mock the views without modifying anything:

```java
        GameScreenTestUtils.setGameUi(gameScreen, mock(ScreenViewport.class));
        GameScreenTestUtils.setGameWorldView(gameScreen, mock(ExtendViewport.class));

```

**Review by TheBromo — CHANGES_REQUESTED** [review]

**Comment by BetaTH13** [comment]
@TheBromo i implemented the Test utils they worked surprisingly well. I alsofixed the issue where the camera was reset

**Review by TheBromo — APPROVED** [review]
lgtm

**Review by BetaTH13 — APPROVED** [review]
